### DISCUSSION TITLE
avoid sending error message as response in `axum`

### DISCUSSION THREAD

**Discussion by zahash** [tokio-rs/axum#2986] [Q&A]
### Summary

if i fail to layer an extension, i get this message in the response body with a 500 error

```
Missing request extension: Extension of type `sqlx_core::pool::Pool<sqlx_sqlite::database::Sqlite>` was not found. Perhaps you forgot to add it? See `axum::Extension`.
```

i instead want to just log it and send an empty response body.

```rust
async fn main() -> Result<(), anyhow::Error> {
    // ...
    let pool = SqlitePool::connect(&database_url).await?;

    let app = Router::new()
        .route("/", get(home))
        // some other routes ...
        .layer(Extension(pool)) // if i remove this line
        ;

    // ...
}
```

### axum version

0.7.6

**Comment by yanns** [comment]
In general, it's now advise to use `State` instead of `Extension`: https://docs.rs/axum/latest/axum/#sharing-state-with-handlers
That way, you will get a compilation error if you forget to set the state.

See https://docs.rs/axum/latest/axum/#using-request-extensions
> The downside to this approach is that you’ll get runtime errors (specifically a 500 Internal Server Error response) if you try and extract an extension that doesn’t exist, perhaps because you forgot to add the middleware or because you’re extracting the wrong type.

**Reply by zahash** [reply]
what is a rule of thumb on when to use state and when to use extractors?

**Reply by zahash** [reply]
@yanns 

i also have a middleware that assigns a unique request id to each request that the handlers can access as RequestId

```rust
pub struct RequestId(String);
impl RequestId {
    pub fn new() -> Self {
        Self(Uuid::new_v4().to_string())
    }
}


let app = Router::new()
        .route("/", get(home))
        // some other routes ...
        .layer(Extension(pool))
        .layer(axum::middleware::from_fn(
            |mut request: Request<Body>, next: Next| async {
                let request_id = RequestId::new();
                request.extensions_mut().insert(request_id);
                next.run(request).await
            },
         ));
        

pub async fn foo(
    Extension(request_id): Extension<RequestId>,
    Extension(pool): Extension<SqlitePool>,
) -> Result<StatusCode, HandlerError> { ... }
```

should that be state or extension?

**Reply by yanns** [reply]
state was introduced after extension.
You should always prefer state over extension as it provides more guarantees at compilation time.

**Reply by jplatte** [reply]
`State` is for application state. When extracting state for different requests, you always get copies of (parts of) the same underlying state struct.

So, use state for application-level resources like database pools.

Use request extensions for request-specific data, like request IDs or authentication information derived from cookies or headers.

**Comment by mladedav** [comment]
You can use `Option<Extension<SqlitePool>>`, then you can react to the missing extensions. But in this case you really do want to use `State` instead, as noted before.

You could also try using [`WithRejection`](https://docs.rs/axum-extra/latest/axum_extra/extract/struct.WithRejection.html) to directly handle the missing extension and change the behavior to what you've describe.
